sklearn/cluster/tests/test_k_means.py

```diff
@@ -85,16 +149,47 @@ def test_elkan_results(distribution, tol):
 
     km_full.fit(X)
     km_elkan.fit(X)
-    assert_array_almost_equal(km_elkan.cluster_centers_,
-                              km_full.cluster_centers_)
+    assert_allclose(km_elkan.cluster_centers_, km_full.cluster_centers_)
     assert_array_equal(km_elkan.labels_, km_full.labels_)
 
-    # The number of iterations and inertia should be close but not
-    # necessarily exactly the same because of rounding errors.
-    assert km_elkan.n_iter_ == pytest.approx(km_full.n_iter_, rel=0.01)
+    assert km_elkan.n_iter_ == km_full.n_iter_
     assert km_elkan.inertia_ == pytest.approx(km_full.inertia_, rel=1e-6)
 
 
+@pytest.mark.parametrize('algorithm', ['full', 'elkan'])
+def test_kmeans_convergence(algorithm):
+    # Check that KMeans stops when convergence is reached when tol=0. (#16075)
+    rnd = np.random.RandomState(0)
+    X = rnd.normal(size=(5000, 10))
+
+    km = KMeans(algorithm=algorithm, n_clusters=5, random_state=0, n_init=1,
+                tol=0, max_iter=300).fit(X)
+
+    assert km.n_iter_ < 300
+
+
+@pytest.mark.parametrize('distribution', ['normal', 'blobs'])
```

Comment:
Aren't the blobs Gaussian too?

Reply:
Yes, but normal here is just 1 gaussian.
This parametrization was already there. It allows to test structured dataset vs unstructured dataset.